Add OTP feature

piyush-gangrade opened the pull request:
PR #39

Add verifyOTP.jsx for sending and verifying OTP, and update Login.jsx and Signup.jsx accordingly.

Details: After users log in or sign up, they will navigate to the 'verify' route where they have to enter the OTP sent to their email. After submitting the OTP, if it is valid, a token will be sent to the user.


![Screenshot (60)](https://github.com/TechNodes2-0/ElectiveHub/assets/133166222/6cce2fc6-ebe8-45fd-a5ff-5ef4cdfe2f38)

piyush-gangrade committed:
add email sent functionality

piyush-gangrade committed:
update UserModel.js and add send-token and verify-token router

add two new fields otpToken and otpExpiry in UserModel.js and two controller function for sending and verifying otp

piyush-gangrade committed:
add verify route and update Login.jsx and Signup.jsx

Add VerifyOTP.jsx and update Login.jsx and SignUp.jsx so that after login and signup, the user navigates to the verify route, and an OTP is sent to the user's email. After entering the OTP, the user will be logged in

JayeshYadav99 commented:
Send screen recording of running this pr on local and performing otp functionality

piyush-gangrade commented:
@JayeshYadav99 
[Uploading OTP feature video.zip…]()

JayeshYadav99 commented:
@piyush-gangrade  Remove otp  verification from login only keep it in signup

piyush-gangrade committed:
remove otp verification from login

piyush-gangrade commented:
@JayeshYadav99, the following changes have been made

VinayakVispute commented:
@piyush-gangrade Hello,

Your contribution looks good to me but there are 2 huge flows

1. There is loophole that immediate attention. Users should not be able to bypass OTP verification simply not verifying the OPT as  signup Page and logining in directly. As you can see in Video.

`Keep user authentication data and OTPs in separate collections to prevent bypassing OTP verification during login.`

https://github.com/TechNodes2-0/ElectiveHub/assets/93467074/acf663fa-af06-4de0-930d-96020405fc4a




2. You OPT are stored in plain text in mongodb. So I think it should encrypted but its your take what you think is correct



![image](https://github.com/TechNodes2-0/ElectiveHub/assets/93467074/268693be-da7b-4272-a8cc-dcbc9bf2b17f)

piyush-gangrade commented:
ok, I'll fix this issue and update

piyush-gangrade committed:
change otp verification functionality

update Signup.jsx, VeifyOTP.jsx, AuthController.js and AuthRoutes and add OtpModel.js and encrypt.js

piyush-gangrade commented:
Changes:

Additions:

1. Added server/Models/OtpModel.js to save email, signature, publicKey, and createdAt.
2. Added server/util/encrypt.js to generate privateKey and publicKey so that we can create a signature of data using privateKey and verify it using publicKey.

Updates:

1. Updated Signup.jsx so that it can send an OTP and provide inputData to VerifyOTP.jsx with the help of the navigate state and location state.
2. Updated the sendOtp function of AuthController.js to create an OTP document, save email, signature, and publicKey, and send an OTP to the given email.
3. Updated VerifyOTP.jsx so that after submitting the OTP, it sends a request to the 'signup' API route with inputData and OTP.
4. Updated the signup function of AuthController.js so that it first checks if the OTP is valid. If the OTP is valid, it allows the creation of a User and sends back a token


Uploading Change OTP feature functionality.mp4…

piyush-gangrade commented:
Thank you! Could you also please add assignees and labels
